### PR TITLE
NIFI-8161 NiFi EL: migration from SimpleDateFormat to DateTimeFormatter: rebased to 2.0

### PR DESCRIPTION
**Pull request by arkadius**
# Summary

Optimization improvement: Migrates NiFi Expression Language from SimpleDateFormat to DateTimeFormatter
[NIFI-8161](https://issues.apache.org/jira/browse/NIFI-8161). This pull request is "reopen" of #4773 after rebase to 2.0, main branch

# Tracking

Please complete the following tracking steps prior to pull request creation.

### Issue Tracking

- [x] [Apache NiFi Jira](https://issues.apache.org/jira/browse/NIFI) issue created

### Pull Request Tracking

- [x] Pull Request title starts with Apache NiFi Jira issue number, such as `NIFI-00000`
- [x] Pull Request commit message starts with Apache NiFi Jira issue number, as such `NIFI-00000`

### Pull Request Formatting

- [x] Pull Request based on current revision of the `main` branch
- [  ] Pull Request refers to a feature branch with one commit containing changes

# Verification

Please indicate the verification steps performed prior to pull request creation.

### Build

- [ ] Build completed using `mvn clean install -P contrib-check`
  - [x] JDK 11
  - [ ] JDK 17

### Licensing

- [x] New dependencies are compatible with the [Apache License 2.0](https://apache.org/licenses/LICENSE-2.0) according to the [License Policy](https://www.apache.org/legal/resolved.html)
- [x] New dependencies are documented in applicable `LICENSE` and `NOTICE` files

### Documentation

- [x] Documentation formatting appears as expected in rendered files

### CODE REVIEW

**Comment by arkadius** [comment]
I rebased this branch as we discussed in the original PR, but I see that meantime NiFi code become much more dependent on `DateFormat`. I see many public methods in `DataTypeUtils` taking or returning `DateFormat`. I see some transition methods like `DateTypeUtils.getDateTimeFormatter`. It makes me feel that:
- I don't see the wider picture of where you already use `DateFormat` now and how you planning to use it in the future
- I don't know what is the plan to get rid of `DateFormat` because it looks like there is some
- I don't know if you want to use a lenient, case insensitive format in all these places or some other one

If you want to draw me a wider picture I can try to continue this change but otherwise, I don't want to introduce some ambiguous behaviour in date format handling so I'll close this PR.

**Comment by MikeThomsen** [comment]
@arkadius you raise some good questions. I would recommend not closing the PR. I'll raise the issue on the developer mailing list (which you are free to join, as it's public).

**Comment by exceptionfactory** [comment]
Thanks for raising these important questions @arkadius! I [replied to the developer list thread](https://lists.apache.org/thread/7j3t58mc9v567ms97zv4dgm3rwshp78k) that @MikeThomsen started.

We want to move in the direction of using the `java.time` classes across the framework, but it will require careful orchestration.

From that perspective, I think keeping this particular PR focused on the Expression Language function is a good starting place, and we can consider the broader impact in separate issues, related to Jira epic [NIFI-9458](https://issues.apache.org/jira/browse/NIFI-9458).

I will take a closer look at this PR soon, but I think it is generally in a good position to review.

**Comment by MikeThomsen** [comment]
@arkadius are you at a point where you want a review or do you want to keep going?

**Comment by arkadius** [comment]
It's ready. I forgot to change the status. I didn't introduce any changes since last review. I only rebased the branch

File: nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/StringToDateEvaluator.java [hunk 1]
```diff
@@ -17,27 +17,51 @@
...
+    private final DateTimeFormatter preparedFormatter;
+    private final boolean preparedFormatterHasRequestedTimeZone;
+
     public StringToDateEvaluator(final Evaluator<String> subject, final Evaluator<String> format, final Evaluator<String> timeZone) {
         this.subject = subject;
         this.format = format;
+        // if the search string is a literal, we don't need to prepare formatter each time; we can just
+        // prepare it once. Otherwise, it must be prepared for each time.
+        if (format instanceof StringLiteralEvaluator) {
+            DateTimeFormatter dtf = FormatUtils.prepareLenientCaseInsensitiveDateTimeFormatter(format.evaluate(new StandardEvaluationContext(Collections.emptyMap())).getValue());
```
Comment (exceptionfactory) [inline nifi-commons/nifi-expression-language/src/main/java/org/apache/nifi/attribute/expression/language/evaluation/functions/StringToDateEvaluator.java:50]: See above note on nested calls and separate variable declaration.

**Review by exceptionfactory — APPROVED** [review]
Thanks again for working on these changes @arkadius. The general approach looks good, I just noted some minor organization adjustments to make the code easier to follow. Other than that, it looks close to completion.

**Comment by arkadius** [comment]
Thank you for the review @exceptionfactory . I applied all suggestions and changed the code as you proposed.

**Review by exceptionfactory — APPROVED** [review]
Thanks for reviving this pull request and making the changes @arkadius. The latest version looks good and works as expected.

The changes in formatting characters from [SimpleDateFormat](https://docs.oracle.com/javase/8/docs/api/java/text/SimpleDateFormat.html) to [DateTimeFormatter](https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html) will be important to highlight in migration documentation for NiFi 2.0, but that applies to a number of changes for the major release version.

+1 merging

**Pull request closed without merging**